### ISSUE TITLE
change partitioning logic from absolutes to a rotating permanent partitioning

### ISSUE THREAD

**Issue by zebastian** [tango-controls-hdbpp/libhdbpp-mysql#15]
The mysql table partitions are hardcoded here and outdated:
libhdbpp-mysql/etc/create_hdb++_mysql_innodb.sql 

for example:
```
CREATE TABLE IF NOT EXISTS att_scalar_devboolean_rw
(
...
) ENGINE=InnoDB COMMENT='Scalar Boolean ReadWrite Values Table'
PARTITION BY RANGE COLUMNS(data_time) (
	PARTITION p000 VALUES LESS THAN ('2015-01-01'),
        PARTITION p2015 VALUES LESS THAN ('2016-01-01'),
        PARTITION p2016 VALUES LESS THAN ('2017-01-01'),
        PARTITION p2017 VALUES LESS THAN ('2018-01-01'),
        PARTITION p2018 VALUES LESS THAN ('2019-01-01'),
        PARTITION future       VALUES LESS THAN MAXVALUE
);
```

Instead a rotating partitioning could be used, which:
- does not need any modifications or updates over time
- still partitions the data in a yearly partition

This can be achieved by:
```
CREATE TABLE IF NOT EXISTS att_scalar_devboolean_rw
(
...
) ENGINE=InnoDB COMMENT='Scalar Boolean ReadWrite Values Table'
PARTITION BY HASH(YEAR(data_time)) PARTITIONS 100;
```

This will create 100 partitions, 
att_scalar_devboolean_rw#p#p0.ibd till att_scalar_devboolean_rw#p#p99.ibd
where the hashing currently assigns the att_scalar_devboolean_rw#p#p21.ibd as the current parttion. 
Confirmed working on mariadb and followed on mysql documentation.

Note that the parttions rotate by the hash function:
HASH = MOD(expr, num)
where num is 100 as set above and expr is YEAR(data_time).
eg: 2021 and 2121 will both write to att_scalar_devboolean_rw#p#p21.ibd

The value of 100 is chosen by personal preference.

see also more info here: https://dev.mysql.com/doc/refman/5.7/en/partitioning-hash.html

Please let me know, what you think of this. IMHO this simplifies usage and documentation a good deal.

I can also help with the code / documentation changes.

Cheers! Sebastian

**Comment by zebastian** [comment]
by the way: is there any measurable performance gains by using the partitioning and not using a plain innodb table instead?

**Comment by gscalam** [comment]
In the hdbpp-mysql-project -> resources -> schema (https://github.com/tango-controls-hdbpp/hdbpp-mysql-project/tree/master/resources/schema) I adopted a different approach:
Schema script is plain, without partitioning, partitioning script takes the current year and creates partition for 2 years, which is actually our configuration at Elettra.
Partitioning benefits depends on different aspects, but mainly it's is matter of DB maintenance, not performance.
In fact we need partitioning to move 3 years old and older partition to a separate DB and keep the running DB limited in size.
You can have an improvement in performance if your queries need to scan the full table (partition) and that doesn't fit in memory.
Using InnoDB, with clustered primary key, querying by att_conf_id and data_time is never an issue, and can only be slightly slower with partitioning due to scan of partition list. Data are clustered on disk by primary key, so, in a way, already partitioned.
Examples of queries which need to scan the whole table (partition) are when you don't specify the attribute, so you're looking for more recent values between all attributes, or the max/min values between all attributes, etc.
If you use different engines, like MyIsam, the situation is different.

Furthermore partitioning is essential for maintenance operations like moving/deleting data, updating schema, etc., when your whole table doesn't fit in memory and instead your partition do.

For the above reasons I don't think in general partitioning by hash is a good choice.

In conclusion I think in the libhdbpp-mysql repo, schema should be removed, and documented that hdbpp-mysql-project should be used. In that repo we can leave schema without partitioning and add (and document) as many partitioning scripts as needed.

**Comment by zebastian** [comment]
Hi @gscalamera ,
thanks for the detailed picture. 
IMHO in the umbrella project "hdbpp-mysql-project" the default schema is fine and renders this ticket closed. Arguably, when you take care of manually deleting partitions this should be a manual or semi manual process and does not need the rotation.

I agree, the libhdbpp-mysql should at least mark the DDL folder as legacy and point to the hdbpp-mysql-project instead.
